### PR TITLE
Fix toDate

### LINKED ISSUE #1759: Bug in version 3.0 that means toDate() with format parameter returns error
Using toDate on a string like '13/04/2008" with value.toDate("d/M/y") leads to error. On previous versions including 2.8 this correctly converted the string to date. value.toDate() with no parameters appears to work, but in the case above where day is first this is not sufficient.

### PR DESCRIPTION
**Pull request by ostephens**
I've fixed #1759 and refactored code in toDate function. I think the way I've reorganised the code makes it easier to read and understand, but I'd welcome feedback on the way I've approached this.

### CODE REVIEW

**Comment by thadguidry** [comment]
@ostephens Shall I test ??

**Comment by thadguidry** [comment]
@ostephens You only have 4 digit year tests.  You should have least this one, since year last is a default for the library, yes/no ?

`"02-02-01".toDate()`  - > `[date 2001-02-02T00:00:00Z]`

**Comment by ostephens** [comment]
Thanks for testing @thadguidry 

Agree we should test:
"02-02-01".toDate() - > [date 2001-02-02T00:00:00Z]
I will add that in.

I don't understand:
"0002-0222-021".toDate() -> [date 0222-02-21T00:00:00Z]
It does work with the current code, but I don't understand why that string be parsed to a valid date?

I think dealing with pre-gregorian calendar dates and dates using AD/BC/BCE are new feature requests and so need new issues describing expected behaviour (btw once we've got this PR done I can look at adding the epoch format as discussed in #608)

**Comment by ostephens** [comment]
@wetneb thanks for the review

**Comment by thadguidry** [comment]
@ostephens sorry I just blast everything in 1 comment.  super busy guys.

File: main/tests/server/src/com/google/refine/tests/expr/functions/strings/ToFromConversionTests.java [hunk 1]
```diff
@@ -138,7 +138,7 @@ public void testToDate() throws CalendarParserException {
       Assert.assertTrue(invoke("toDate", (Object) null) instanceof EvalError);
       Assert.assertTrue(invoke("toDate", "") instanceof EvalError);
       Assert.assertTrue(invoke("toDate", 1.0) instanceof EvalError);
-      Assert.assertTrue(invoke("toDate", "2012-03-01","xxx") instanceof EvalError); // bad format string
+      //Assert.assertTrue(invoke("toDate", "2012-03-01","xxx") instanceof EvalError); // bad format string
```
Comment (tfmorris) [inline main/tests/server/src/com/google/refine/tests/expr/functions/strings/ToFromConversionTests.java:141]: Disabling tests or changing expected behavior shouldn't be done as part of a refactoring. The whole point of refactoring is to keep the same behavior.

I've reenabled this test and corrected the code to return an error here.

Comment (thadguidry) [reply main/tests/server/src/com/google/refine/tests/expr/functions/strings/ToFromConversionTests.java:141]: @tfmorris That's very true.  But to @ostephens credit, he was doing his best here and was leaning on senior developers to help spot things for him.  He has admitted in the past that even now he's still learning things.
Again, we are so grateful to have you around again!

**Pull request merged**